File: impl/ts/tests/timezone.test.ts
// Cross-core time-zone contract: the TS RFC 8536 reader + the JTZ bundle codec must reproduce the
// byte-exact vectors in spec/tz/vectors/{tzif,bundle}.toml (CLAUDE.md §8; spec/tz/README.md §3/§4).
// Mirrors impl/rust/tests/timezone.rs and impl/go/timezone_test.go.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import {
  loadTimeZoneData,
  offsetAtRef,
  openTzBundle,
  resolveZone,
  saveTzBundle,
} from "../src/timezone.ts";
import { readTomlTables, specPath } from "./tomlmini.ts";

function bundleBytes(): Uint8Array {
  // Normalize the Node Buffer to a plain Uint8Array so deepStrictEqual against saveTzBundle's
  // Uint8Array compares bytes, not the Buffer-vs-Uint8Array type.
  return Uint8Array.from(readFileSync(specPath("tz/fixtures/tzdata.jtz")));
}

const SECS = 1_000_000n;

test("timezone reader matches the pinned vectors", () => {
  loadTimeZoneData(bundleBytes());
  const rows = readTomlTables(specPath("tz/vectors/tzif.toml"), "case");
  assert.ok(rows.length > 0, "no tzif vectors");
  for (const row of rows) {
    const zone = row.str("zone");
    const inst = row.big("instant_micros");
    const zr = resolveZone(zone);
    assert.ok(zr !== undefined, `resolve ${zone}`);
    const off = offsetAtRef(zr, inst / SECS - (inst < 0n && inst % SECS !== 0n ? 1n : 0n));
    assert.equal(BigInt(off.utoff), row.big("utoff_secs"), `${zone} @ ${inst}: utoff`);
    assert.equal(off.abbrev, row.str("abbrev"), `${zone} @ ${inst}: abbrev`);
    assert.equal(off.isDst, row.bool("is_dst"), `${zone} @ ${inst}: is_dst`);
  }
});

test("timezone bundle matches the pinned vectors", () => {
  const data = bundleBytes();
  const parsed = openTzBundle(data);
  const rows = readTomlTables(specPath("tz/vectors/bundle.toml"), "bundle");
  assert.equal(rows.length, 1, "want 1 bundle row");
  const b = rows[0];

  assert.equal(parsed.tzdataVersion, b.str("tzdata_version"));
  assert.deepEqual(
    parsed.zones.map((z) => z.name),
    b.strs("zones"),
    "zone manifest",
  );
  assert.deepEqual(
    parsed.links.map((l) => `${l.alias}=${l.target}`),
    b.strs("links"),
    "link table",
  );
  assert.deepEqual(saveTzBundle(parsed), data, "bundle round-trip");
});

File: impl/go/timezone_test.go
package jed

// Cross-core time-zone contract: the Go RFC 8536 reader + the JTZ bundle codec must reproduce the
// byte-exact vectors in spec/tz/vectors/{tzif,bundle}.toml (CLAUDE.md §8; spec/tz/README.md §3/§4).
// Mirrors impl/rust/tests/timezone.rs and impl/ts/tests/timezone.test.ts.

import (
	"bytes"
	"fmt"
	"os"
	"testing"
)

func tzBundleBytes(t *testing.T) []byte {
	t.Helper()
	data, err := os.ReadFile(specPath(t, "tz/fixtures/tzdata.jtz"))
	if err != nil {
		t.Fatalf("read tzdata.jtz: %v", err)
	}
	return data
}

func TestTimezoneReaderMatchesVectors(t *testing.T) {
	if err := LoadTimeZoneData(tzBundleBytes(t)); err != nil {
		t.Fatalf("load tzdata.jtz: %v", err)
	}
	rows := readTomlTables(t, specPath(t, "tz/vectors/tzif.toml"), "case")
	if len(rows) == 0 {
		t.Fatal("no tzif vectors")
	}
	for _, row := range rows {
		zone := row.str("zone")
		inst := row.int("instant_micros")
		zr, ok := ResolveZone(zone)
		if !ok {
			t.Fatalf("resolve %s", zone)
		}
		off := offsetAtRef(zr, floorDiv(inst, 1_000_000))
		if int64(off.Utoff) != row.int("utoff_secs") {
			t.Errorf("%s @ %d: utoff = %d, want %d", zone, inst, off.Utoff, row.int("utoff_secs"))
		}
		if off.Abbrev != row.str("abbrev") {
			t.Errorf("%s @ %d: abbrev = %q, want %q", zone, inst, off.Abbrev, row.str("abbrev"))
		}
		if off.IsDst != row.boolVal("is_dst") {
			t.Errorf("%s @ %d: is_dst = %v, want %v", zone, inst, off.IsDst, row.boolVal("is_dst"))
		}
	}
}

func TestTimezoneBundleMatchesVectors(t *testing.T) {
	data := tzBundleBytes(t)
	parsed, err := openTzBundle(data)
	if err != nil {
		t.Fatalf("open tzdata.jtz: %v", err)
	}
	rows := readTomlTables(t, specPath(t, "tz/vectors/bundle.toml"), "bundle")
	if len(rows) != 1 {
		t.Fatalf("want 1 bundle row, got %d", len(rows))
	}
	b := rows[0]

	if parsed.TzdataVersion != b.str("tzdata_version") {
		t.Errorf("tzdata_version = %q, want %q", parsed.TzdataVersion, b.str("tzdata_version"))
	}

	gotZones := make([]string, len(parsed.Zones))
	for i, z := range parsed.Zones {
		gotZones[i] = z.Name
	}
	wantZones := b.strs("zones")
	if fmt.Sprint(gotZones) != fmt.Sprint(wantZones) {
		t.Errorf("zones = %v, want %v", gotZones, wantZones)
	}

	gotLinks := make([]string, len(parsed.Links))
	for i, l := range parsed.Links {
		gotLinks[i] = l.Alias + "=" + l.Target
	}
	wantLinks := b.strs("links")
	if fmt.Sprint(gotLinks) != fmt.Sprint(wantLinks) {
		t.Errorf("links = %v, want %v", gotLinks, wantLinks)
	}

	if !bytes.Equal(saveTzBundle(parsed), data) {
		t.Error("bundle round-trip is not byte-identical")
	}
}
